README.md
<p align="center" >
<img src="https://github.com/particle-iot/particle-cloud-sdk-ios/raw/master/particle-mark.png" alt="Particle" title="Particle">
</p>

# Particle iOS Cloud SDK

[![Build Status](https://travis-ci.org/particle-iot/particle-cloud-sdk-ios.svg?branch=master)](https://travis-ci.org/particle-iot/particle-cloud-sdk-ios)
[![Coverage Status](https://coveralls.io/repos/github/particle-iot/particle-cloud-sdk-ios/badge.svg?branch=master)](https://coveralls.io/github/particle-iot/particle-cloud-sdk-ios?branch=master)

### Introduction


Particle iOS Cloud SDK enables iOS apps to interact with Particle-powered connected products via the Particle Cloud. It’s an easy-to-use wrapper for Particle REST API. The Cloud SDK will allow you to:

- Manage & inject user sessions for the Particle Cloud (access tokens, encrypted session management)
- Claim/Unclaim devices for a user account
- Get a list of Particle devices claimed by user
- Read variables from devices
- Invoke functions on devices
- Publish events from the mobile app and subscribe to events coming from devices
- Get data usage information for Electron devices

### Getting Started

- Download [Particle iOS Cloud SDK](https://github.com/particle/particle-cloud-sdk-ios/archive/master.zip) and try out the included iPhone example app
- Refer to our [documentation](https://docs.particle.io/reference/ios/) for getting started guide, API reference, support & feedback links.


### Communication

- If you **need help**, use [Our community website](http://community.particle.io)
- If you **found a bug**, _and can provide steps to reliably reproduce it_, open an issue.
- If you **have a feature request**, open an issue.
- If you **want to contribute**, submit a pull request.

### Release Policy & Cadence

We try to make sure tip of master branch always compiles. We will tag commits used for AppStore releases with git tag. Any time something is merged to master, we will ship a release which includes that change within 90 days.

(Note: while it is our intent to release on this schedule, it is not meant as a guarantee. Circumstances may require us to expand the window between a change and when that change ships in a release.)

### License

Particle iOS Cloud SDK is available under the Apache License 2.0. See the [LICENSE file](https://github.com/particle-iot/particle-cloud-sdk-ios/blob/master/LICENSE) for more info.


CHANGELOG.md
#Change Log
All notable changes to this project will be documented in this file.
Particle iOS Cloud SDK adheres to [Semantic Versioning](http://semver.org/).

---
## [1.0.7](https://github.com/particle-iot/particle-cloud-sdk-ios/releases/tag/1.0.7) (2021-04-24)

* Fixing an issue with cocoapods

## [1.0.6](https://github.com/particle-iot/particle-cloud-sdk-ios/releases/tag/1.0.6) (2021-02-23)

* Added support for developer plans

## [1.0.5](https://github.com/particle-iot/particle-cloud-sdk-ios/releases/tag/1.0.5) (2020-04-24)

* Update AFNetworking to 4.0 to fix AppStore error about usage of deprecated UIWebView

## [1.0.4](https://github.com/particle-iot/particle-cloud-sdk-ios/releases/tag/1.0.4) (2020-03-23)

* Add custom header to easier identify traffic coming from mobile SDK

## [1.0.3](https://github.com/particle-iot/particle-cloud-sdk-ios/releases/tag/1.0.3) (2020-02-04)

* Added support for platform id 25 (B5 SoM / BSoM CAT 1)

* Switched to using AFNetworking subspecs to silence appstore warning shown when submitting an app update

## [1.0.2](https://github.com/particle-iot/particle-cloud-sdk-ios/releases/tag/1.0.2) (2020-01-13)

* Added TravisCI integration and few unit tests

## [1.0.1](https://github.com/particle-iot/particle-cloud-sdk-ios/releases/tag/1.0.1) (2019-11-08)

* Changed name of the repository, podspec & readme file.

## [1.0.0](https://github.com/particle-iot/particle-cloud-sdk-ios/releases/tag/1.0.0) (2019-11-04)

* Changed access token logging behavior to always log just last 4 digits

* Fixed a bug causing app to crash when attempting to resume session from the app running in the background

* Removed `requestPasswordResetForCustomer` as it is no longer supported by Particle Cloud API

## [0.9.7](https://github.com/particle-iot/particle-cloud-sdk-ios/releases/tag/0.9.7) (2019-07-30)

* Removed unused `networkState` property from `ParticleDevice`

* Added support for `networkId`, `networkRole`, `networkRoleState` (device roles need to be approved) properties

## [0.9.6](https://github.com/particle-iot/particle-cloud-sdk-ios/releases/tag/0.9.6) (2019-06-27)

* Added support for editing device notes and pinging the device

* Added support for custom ParticleCloud API base URL

* Added support for `ParticleDeviceSystemEventFlashFailed` event

!!! Potentially breaking change !!!
* Preciusly unused `version` property has been renamed to `systemFirmwareVersion` and now it contains DeviceOS version string.

## [0.9.5](https://github.com/particle-iot/particle-cloud-sdk-ios/releases/tag/0.9.5) (2019-04-12)

* Added support for endpoints related to SIM & credit card management used in 3rg generation device guided setup. 

## [0.9.4](https://github.com/particle-iot/particle-cloud-sdk-ios/releases/tag/0.9.4) (2019-03-05)

!!! Potentially breaking change !!!
* Call to `ParticleCloud.getDevices()` will no longer call `ParticleCloud.getDevice()` for every device that is reported as being online. To receive full device information, call `ParticleCloud.getDevice()` or `ParticleDevice.refresh()` on a device instance returned by `ParticleCloud.getDevices()`

## [0.9.3](https://github.com/particle-iot/particle-cloud-sdk-ios/releases/tag/0.9.3) (2019-02-25)

* Introduced 4th log level `Complete` that outputs very noisy data sources (such as getDevice API call)

## [0.9.2](https://github.com/particle-iot/particle-cloud-sdk-ios/releases/tag/0.9.2) (2019-02-11)

* Increased max length for function argument from 62 chars to 622 chars

## [0.9.1](https://github.com/particle-iot/particle-cloud-sdk-ios/releases/tag/0.9.1) (2019-02-08)

* Added support for 3rd generation SOM devices

* Added support for retrieving device platform ID from serial number

## [0.9.0](https://github.com/particle-iot/particle-cloud-sdk-ios/releases/tag/0.9.0) (2019-01-29)

* Added support for 3rd generation devices

* Fixed a crashing EventStreams bug

## [0.8.1](https://github.com/particle-iot/particle-cloud-sdk-ios/releases/tag/0.8.1) (2018-08-03)

* Bug fixes

## [0.8.0](https://github.com/particle-iot/particle-cloud-sdk-ios/releases/tag/0.8.0) (2018-08-01)

* Added support for two-step verification

* Improved error reporting. `NSError.code` now contains HTTP status code. Added 2 values to userInfo dictionary. `ParticleSDKErrorResponseBodyKey` is the NSDictionary representation of JSON server response. `ParticleSDKErrorLocalizedStringKey` contains human readable error message.  `NSError.localizedDescription` is our best attempt to explain what happened in human readable language based on `ParticleSDKErrorLocalizedStringKey` and `NSError.code`.

* New library folder structure

* New example app

## [0.7.1](https://github.com/particle-iot/particle-cloud-sdk-ios/releases/tag/0.7.1) (2018-06-24)

* BugFix: Replaced outdated and complicated keychain code with simplified version.

## [0.7.0](https://github.com/particle-iot/particle-cloud-sdk-ios/releases/tag/0.7.0) (2017-04-04)

* Overdue rename from particle-iot to Particle! particle-iotCloud is now ParticleCloud, same for ParticleDevice etc.

* Pod has been renamed from particle-iot-SDK to Particle-SDK. particle-iot-SDK pod has been deprecated.

* Documentation update

## [0.6.0](https://github.com/particle-iot/particle-cloud-sdk-ios/releases/tag/0.6.0) (2017-02-03)

* Updated: Product endpoints deprecates Org endpoints

* Added: Raspberry pi platform 31 support

* Added: accountInfo field to createUser (add optional first/last name, company name and business account info to every user signup)

* Bigfix: ParticleDevice.refresh() now does not nullify the delegate property

* Bigfix: Internal device maptable issue

* Bigfix: System events subscribe only when SDK has an active access token

* Added: Syntax changes for interoperability with Xcode 8 / Swift 3.0

* Bigfix: MBs Usage for Electron endpoint API response fix

## [0.5.1](https://github.com/particle-iot/particle-cloud-sdk-ios/releases/tag/0.5.1) (2016-09-07)


ParticleSDK/SDK/ParticleDevice.h
//
//  ParticleDevice.h
//  Particle iOS Cloud SDK
//
//  Created by Ido Kleinman
//  Copyright 2015 Particle
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

#import <Foundation/Foundation.h>
#import "ParticleEvent.h"

NS_ASSUME_NONNULL_BEGIN

/**
 * Standard completion block for API calls, will be called when the task is completed
 * with a nullable error object that will be nil if the task was successful.
 */
typedef void (^ParticleCompletionBlock)(NSError * _Nullable error);

typedef NS_ENUM(NSInteger, ParticleDeviceType) {
    ParticleDeviceTypeUnknown=-1,
    ParticleDeviceTypeCore=0,
    ParticleDeviceTypePhoton=6, // or P0
    ParticleDeviceTypeP1=8,
    ParticleDeviceTypeElectron=10, // or E0
    ParticleDeviceTypeRaspberryPi=31,
    ParticleDeviceTypeRedBearDuo=88,
    ParticleDeviceTypeBluz=103,
    ParticleDeviceTypeDigistumpOak=82,
    ParticleDeviceTypeESP32=11,
    ParticleDeviceTypeArgon=12,
    ParticleDeviceTypeBoron=13,
    ParticleDeviceTypeXenon=14,
    ParticleDeviceTypeASeries=22,
    ParticleDeviceTypeBSeries=23,
    ParticleDeviceTypeXSeries=24,
    ParticleDeviceTypeBSoMCat1=25
};

typedef NS_ENUM(NSInteger, ParticleDeviceSystemEvent) {
    ParticleDeviceSystemEventCameOnline,
    ParticleDeviceSystemEventWentOffline,
    ParticleDeviceSystemEventFlashStarted,
    ParticleDeviceSystemEventFlashSucceeded,
    ParticleDeviceSystemEventFlashFailed,
    ParticleDeviceSystemEventAppHashUpdated,
    ParticleDeviceSystemEventEnteredSafeMode,
    ParticleDeviceSystemEventSafeModeUpdater
};

typedef NS_ENUM(NSInteger, ParticleDeviceNetworkRole) {
    ParticleDeviceNetworkRoleGateway,
    ParticleDeviceNetworkRoleNode,
};

typedef NS_ENUM(NSInteger, ParticleDeviceNetworkRoleState) {
    ParticleDeviceNetworkRoleStatePending,
    ParticleDeviceNetworkRoleStatePendingConfirmed,
    ParticleDeviceNetworkRoleStatePendingOwnerApproval
};

@class ParticleDevice;

@protocol ParticleDeviceDelegate <NSObject>

@optional
-(void)particleDevice:(ParticleDevice *)device didReceiveSystemEvent:(ParticleDeviceSystemEvent)event;

@end

@interface ParticleDevice : NSObject

/**
 *  DeviceID string
 */
@property (strong, nonatomic, readonly) NSString* id;

/**
 *  Device name. Device can be renamed in the cloud by setting this property. If renaming fails name will stay the same.
 */
@property (strong, nullable, nonatomic) NSString* name;

/**
 *  Is device connected to the cloud? Best effort - May not accurate reflect true state.
 */
@property (nonatomic, readonly) BOOL connected;

/**
 *  List of function names exposed by device. Functions are only available after `refresh` is called on the device instance returned by `getDevices` call.
 */
@property (strong, nonatomic, nonnull, readonly) NSArray<NSString *> *functions;

/**
 *  Dictionary of exposed variables on device with their respective types. Variables are only available after `refresh` is called on the device instance returned by `getDevices` call.
 */
@property (strong, nonatomic, nonnull, readonly) NSDictionary<NSString *, NSString *> *variables; // @{varName : varType, ...}

@property (strong, nonatomic, nullable, readonly) NSString *lastApp; __deprecated_msg("This property is not updated");

/**
 * Date when the device was last heard.
 */
@property (strong, nonatomic, nullable, readonly) NSDate *lastHeard;

/**
 * App hash received from system event after flashing a new different user application. Please note that if you flashe same app, the app hash event won't trigger.
 */
@property (strong, nonatomic, nullable, readonly) NSString *appHash;

/**
 * Mesh network ID if device belongs to a mesh network. `nil` otherwise.
 */
@property (strong, nonatomic, nullable, readonly) NSString *networkId;

/**
 * Mesh network role if device belongs to a mesh network (gateway or a node).
 */
@property (nonatomic, readonly) ParticleDeviceNetworkRole networkRole;

/**
 * Mesh network state if device belongs to a mesh network. State can be pending if device is waiting for the role change confirmation.
 */
@property (nonatomic, readonly) ParticleDeviceNetworkRoleState networkRoleState;

/**
 * Is device currently being flashed? Best effort - May not accurate reflect true state.
 */
@property (nonatomic, readonly) BOOL isFlashing;

/**
 * Last know IP address of the device.
 */
@property (strong, nonatomic, nullable, readonly) NSString *lastIPAdress;

/**
 * Last known ICCID (for cellular devices only)
 */
@property (strong, nonatomic, nullable, readonly) NSString *lastIccid;

/**
 * Last know IMEI (for cellular devices only)
 */
@property (strong, nonatomic, nullable, readonly) NSString *imei;

/**
 * Device platfomr ID (device type).
 */
@property (nonatomic, readonly) NSUInteger platformId;

/**
 * ID of the product device belongs to.
 */
@property (nonatomic, readonly) NSUInteger productId;


@property (strong, nonatomic, nullable, readonly) NSString *status; __deprecated_msg("This property is not updated");


/**
 * Unique serial number of the device
 */
@property (strong, nonatomic, nullable, readonly) NSString *serialNumber;

/**
 * Mobile secret of the device. Only available for 3rd gen devices.
 */
@property (strong, nonatomic, nullable, readonly) NSString *mobileSecret;

/**
 * Additional device notes. User editable.
 */
@property (strong, nonatomic, nullable) NSString *notes;

/**
 *  Device firmware version string
 */
@property (strong, nonatomic, readonly) NSString *systemFirmwareVersion;

/**
 * Is this device a cellular device?
 */
@property (nonatomic, readonly) BOOL cellular;

/**
 * Does this device require software update?
 */
@property (nonatomic, readonly) BOOL requiresUpdate;

/**
 * Particle device type. A easier to use version of `platformId`.
 */
@property (nonatomic, readonly) ParticleDeviceType type;

/**
 * `type` casted into a human readable String.
 */
@property (nonatomic, readonly) NSString *typeString;



-(nullable instancetype)initWithParams:(NSDictionary *)params NS_DESIGNATED_INITIALIZER;
-(instancetype)init __attribute__((unavailable("Must use initWithParams:")));

@property (nonatomic, strong) id <ParticleDeviceDelegate> delegate;

/**
 *  Retrieve a variable value from the device
 *
 *  @param variableName Variable name
 *  @param completion   Completion block to be called when function completes with the variable value retrieved (as id/AnyObject) or NSError object in case on an error
 */
-(NSURLSessionDataTask *)getVariable:(NSString *)variableName completion:(nullable void(^)(id _Nullable result, NSError* _Nullable error))completion;

/**
 *  Call a function on the device
 *
 *  @param functionName Function name
 *  @param args         Array of arguments to pass to the function on the device. Arguments will be converted to string maximum length 63 chars.
 *  @param completion   Completion block will be called when function was invoked on device. First argument of block is the integer return value of the function, second is NSError object in case of an error invoking the function
 */
-(NSURLSessionDataTask *)callFunction:(NSString *)functionName
                        withArguments:(nullable NSArray *)args
                           completion:(nullable void (^)(NSNumber * _Nullable result, NSError * _Nullable error))completion;

/**
 *  Signal device
 *  Will make the onboard LED "shout rainbows" for easy physical identification of a device
 *
 *  @param enable - YES to start or NO to stop LED signal.
 *
 */
-(NSURLSessionDataTask *)signal:(BOOL)enable completion:(nullable ParticleCompletionBlock)completion;


/**
 *  This will ping a device, enabling you to see if your device is online or offline
 */
- (NSURLSessionDataTask *)ping:(nullable void (^)(BOOL result, NSError *_Nullable error))completion;

/**
 *  Request device refresh from cloud
 *  update online status/functions/variables/device name, etc
 *
 *  @param completion Completion block called when function completes with NSError object in case of an error or nil if success.
 *
 */
-(NSURLSessionDataTask *)refresh:(nullable ParticleCompletionBlock)completion;

/**
 *  Remove device from current logged in user account
 *
 *  @param completion Completion block called when function completes with NSError object in case of an error or nil if success.
 */
-(NSURLSessionDataTask *)unclaim:(nullable ParticleCompletionBlock)completion;


/**
 *  Rename device
 *
 *  @param newName      New device name
 *  @param completion   Completion block called when function completes with NSError object in case of an error or nil if success.
 */
-(NSURLSessionDataTask *)rename:(NSString *)newName completion:(nullable ParticleCompletionBlock)completion;

/**
 *  Edit device notes
 *
 *  @param notes        Notes to be associated with particular device
 *  @param completion   Completion block called when function completes with NSError object in case of an error or nil if success.
 */
-(NSURLSessionDataTask *)setNotes:(NSString *)notes completion:(nullable ParticleCompletionBlock)completion;

/**
 *  Retrieve current data usage report (For Electron only)
 *
 *  @param completion   Completion block to be called when function completes with the data used in current payment period in (float)MBs. All devices other than Electron will return an error with -1 value
 */
-(NSURLSessionDataTask *)getCurrentDataUsage:(nullable void(^)(float dataUsed, NSError* _Nullable error))completion;


/**
 *  Flash files to device
 *
 *  @param filesDict    files dictionary in the following format: @{@"filename.bin" : <NSData>, ...} - that is a NSString filename as key and NSData blob as value. More than one file can be flashed. Data is alway binary.
 *  @param completion   Completion block called when function completes with NSError object in case of an error or nil if success. NSError.localized descripion will contain a detailed error report in case of a
 */
-(nullable NSURLSessionDataTask *)flashFiles:(NSDictionary *)filesDict completion:(nullable ParticleCompletionBlock)completion; //@{@"<filename>" : NSData, ...}

/**
 *  Flash known firmware images to device
 *
 *  @param knownAppName    NSString of known app name. Currently @"tinker" is supported. 
 *  @param completion      Completion block called when function completes with NSError object in case of an error or nil if success. NSError.localized descripion will contain a detailed error report in case of a
 */
-(NSURLSessionDataTask *)flashKnownApp:(NSString *)knownAppName completion:(nullable ParticleCompletionBlock)completion; // knownAppName = @"tinker", @"blinky", ... see http://docs.

#pragma mark Events subsystem functions

/**
 *  Subscribe to events from this specific (claimed) device - both public and private.
 *
 *  @param eventNamePrefix  Filter only events that match name eventNamePrefix, for exact match pass whole string, if nil/empty string is passed any event will trigger eventHandler
 *  @param eventHandler     Event handler function that accepts the event payload dictionary and an NSError object in case of an error
 */
-(nullable id)subscribeToEventsWithPrefix:(nullable NSString *)eventNamePrefix handler:(nullable ParticleEventHandler)eventHandler;

/**
 *  Unsubscribe from event/events.
 *
 *  @param eventListenerID The eventListener registration unique ID returned by the subscribe method which you want to cancel
 */
-(void)unsubscribeFromEventWithID:(id)eventListenerID;

// Internal use
-(void)__receivedSystemEvent:(ParticleEvent *)event;

@end

NS_ASSUME_NONNULL_END
